onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java

```diff
@@ -111,33 +111,34 @@ public TaskContextImpl(int startId) {
         public void taskComplete() {
             //Log.d(TAG, "Task complete: " + mStartId);
             // If we have notifications, then we can't stop ourselves.
-            if (mNotifications.isEmpty()) {
+            //TODO: Is there actually a reason to not stop ourselves if we are no longer storing notifications?
```

Comment:
Hi Sean,

It looks to me like all of the remaining processes being executed in
TripService are being executed from the threadpool object
(ExecutorService). I think the key to shutting down safely (if there is
one) is in something like what is outlined here:

http://stackoverflow.com/questions/3269445/executorservice-how-to-wait-for-all-tasks-to-finish

It seems to me that the safest way to wait for a shutdown is to check to
see if the executor is still running some thread, if there is a reasonable
way to determine this.

On Tue, May 10, 2016 at 10:21 AM, Sean Barbeau <notifications@github.com>
wrote:

> In
> onebusaway-android/src/main/java/org/onebusaway/android/tripservice/TripService.java
> <https://github.com/OneBusAway/onebusaway-android/pull/479#discussion_r62714004>
> :
>
> > @@ -111,33 +111,34 @@ public TaskContextImpl(int startId) {
> >          public void taskComplete() {
> >              //Log.d(TAG, "Task complete: " + mStartId);
> >              // If we have notifications, then we can't stop ourselves.
> > -            if (mNotifications.isEmpty()) {
> > +            //TODO: Is there actually a reason to not stop ourselves if we are no longer storing notifications?
>
> Seems like the simplest solution without a complete overahul of the code
> (e.g. to use something like an IntentService) may be to keep the
> mNotifications object, but only use it as a means to ensure we don't shut
> down the service prematurely. SetLatestEventInfo() should still be
> eliminated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/OneBusAway/onebusaway-android/pull/479/files/6352fbd746b5be92d461464ffe018ac52552ef7a#r62714004>
>